httpreplay/cobweb.py

```diff
@@ -175,3 +180,245 @@ def handle(self, s, ts, protocol, sent, recv):
+
+    def handle_mailbody(self, data):
+        headers_mes = data.split("\r\n\r", 1)
+        if len(headers_mes) < 2:
+            return
+
+        self.request.message = headers_mes[1]
+
+        for header in headers_mes[0].split("\r\n"):
+            if ":" not in header:
+                continue
+
+            key, value = header.split(":", 1)
+            self.request.headers[key] = value
+
+    def handle_auth(self, data):
+        """
+        Determines what kind of authentication type was used
+        and tries to collect the credentials based on the used authentication
+        type
+
+        http://www.samlogic.net/articles/smtp-commands-reference-auth.htm
+        """
+        auth_handlers = {
+            "plain": self.handle_auth_plain,
+            "login": self.handle_auth_login
+        }
+
+        if len(data) < 2:
+            return
+
+        arg_first = data[1].lower()
+        if arg_first not in auth_handlers:
+            return
+
+        self.request.auth_type = arg_first
+
+        if len(data) > 2:
+            arg_second = data[2]
+            handler = auth_handlers[arg_first]
+            handler(arg_second)
+
+    def handle_auth_plain(self, arg):
+        try:
+            user_pass = filter(None, arg.decode("base64").split("\x00"))
+            if len(user_pass) < 2:
+                return
+
+            self.request.username = user_pass[0]
+            self.request.password = user_pass[1]
+        except binascii.Error:
+            return
+
+    def handle_auth_login(self, arg):
+        try:
+            self.request.username = arg.decode("base64")
+        except binascii.Error:
+            return
+
+    def handle_auth_cram_md5(self, arg):
+        user_challange = arg.decode("base64").split(" ", 1)
+        if len(user_challange) < 2:
+            return
+
+        self.request.username = user_challange[0]
+
+    def handle_auth_serv_response(self, data):
+        """
+        If not all credentials were passed as an argument to an auth
+        command, they will be later sent when the server requests them. If
+        that happens, this function extracts them
+        """
+        if len(data) < 1:
+            return
+
+        client_response = data[0]
+
+        if self.request.auth_type == "plain":
+            self.handle_auth_plain(client_response)
+
+        elif self.request.auth_type == "cram-md5":
+            self.handle_auth_cram_md5(client_response)
+
+        elif self.request.auth_type == "login":
+            if "UGFzc3dvcmQ6" in self.message:
+                try:
+                    self.request.password = client_response.decode("base64")
+                except binascii.Error:
+                    return
+            elif "VXNlcm5hbWU6" in self.message:
+                try:
+                    self.request.username = client_response.decode("base64")
+                except binascii.Error:
+                    return
```

Comment:
Maybe also put this in a separate function (as you did with `plain` and `cram-md5`)?